layer/VkLayer_FROG_gamescope_wsi.cpp

```diff
@@ -948,23 +1028,25 @@ namespace GamescopeWSILayer {
         }
       }
 
-      VkResult result = pDispatch->QueuePresentKHR(queue, pPresentInfo);
+      VkResult result = pDispatch->QueuePresentKHR(queue, &presentInfo);
 
-      for (uint32_t i = 0; i < pPresentInfo->swapchainCount; i++) {
-        VkSwapchainKHR swapchain = pPresentInfo->pSwapchains[i];
+      for (uint32_t i = 0; i < presentInfo.swapchainCount; i++) {
+        VkSwapchainKHR swapchain = presentInfo.pSwapchains[i];
 
         auto UpdateSwapchainResult = [&](VkResult newResult) {
-          if (pPresentInfo->pResults && pPresentInfo->pResults[i] >= VK_SUCCESS)
-            pPresentInfo->pResults[i] = newResult;
+          if (presentInfo.pResults && presentInfo.pResults[i] >= VK_SUCCESS)
+            presentInfo.pResults[i] = newResult;
           if (result >= VK_SUCCESS)
             result = newResult;
         };
 
         if (auto gamescopeSwapchain = GamescopeSwapchain::get(swapchain)) {
-
-          if ((limiterOverride == 1 && gamescopeSwapchain->presentMode != VK_PRESENT_MODE_FIFO_KHR) ||
-              (limiterOverride != 1 && gamescopeSwapchain->presentMode != gamescopeSwapchain->originalPresentMode)) {
-              fprintf(stderr, "[Gamescope WSI] Forcing swapchain recreation as frame limiter changed.\n");
+          // If we are a frame limiter aware application like DXVK or VKD3D-Proton, we don't
+          // transparently change their vkQueuePresent to just FIFO modes, we change what is
+          // exposed as supported in order for them to handle presentation latency like they
+          // would as in FIFO mode.
```

Comment:
What about non-frame limiter aware applications? I would assume they still should recreate swapchain and force the present mode accordingly?

Reply:
Non-aware games just get injected with a VkPresentMode with maintenance 1 at present time

Reply:
I didn't see where the layer sets up a compatibility group in swapchain creation. That needs to be there, and layer must query present mode compatibility for surface.